### PR TITLE
add `--interface-priority` and `--net-classid` flags

### PR DESCRIPTION
**Pull request by FengtuWang**
**\- What I did**
Add option `--interface-priority` and `--net-classid` to `docker run` 
and `docker create`

`--interface-priority` is used to set the priority of network traffic, 
corresponding net_prio in cgroup.

`--net-classid` is used to tag network packets with a class identifier, 
corresponding net_cls in cgroup.

This PR depends on docker/engine-api/pull/384

**\- How I did it**
Get option values, and then pass them to containerd. 
Change the runc's verion, for commit bb42f80a86997d8265ee31e2aed33a6670859ac1 in 
https://github.com/opencontainers/runc . If i doesn't do it, runc cann't start up.

**\- How to verify it**
Run integration tests `TestRunWithInterfacePriority` and `TestRunWithClassid`

**\- Description for the changelog**

**\- A picture of a cute animal (not mandatory but encouraged)**

`--interface-priority` is used to set the priority of
network traffic, corresponding net_prio in cgroup.
`--net-classid` is used to tag network packets with a
class identifier, corresponding net_cls in cgroup.

Signed-off-by: Fengtu Wang wangfengtu@huawei.com

### CODE REVIEW

File: runconfig/opts/InterfacePriority.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+func (opt *IfPriorityOpt) Set(val string) error {
+	var value *containertypes.IfPriority
+	if opt.validator != nil {
+		v, err := opt.validator(val)
+		if err != nil {
+			return err
+		}
+		value = v
+	}
+	(opt.values) = append((opt.values), value)
```
Comment (AkihiroSuda) [inline runconfig/opts/InterfacePriority.go]: I think you don't need parenthesis here

File: runconfig/opts/InterfacePriority.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	if len(split) != 2 {
+		return nil, fmt.Errorf("bad format: %s", val)
+	}
+
+	prio, err := strconv.ParseUint(split[1], 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("invalid priority for network interface: %s. The correct format is <network-interface>:<number>. Number must be a positive integer", val)
+	}
+	if prio < 0 {
+		return nil, fmt.Errorf("invalid priority for network interface: %s. The correct format is <network-interface>:<number>. Number must be a positive integer", val)
```
Comment (AkihiroSuda) [inline runconfig/opts/InterfacePriority.go]: `fmt.Errorf` code clone?

**Comment by justincormack** [comment]
cc @mrjana

**Comment by FengtuWang** [comment]
changelist: 
1. Change `p` to `priority` to avoid confuse, in file daemon/daemon_unix.go
2. Strip unnecessary parenthesis in file runconfig/opts/InterfacePriority.go
3. Merge two error handling codes to one in file runconfig/opts/InterfacePriority.go.
   Corresponding commit `fmt.Errorf code clone?`
4. Add prefix to Priorities and ClassID, so variables and docs in some files changed.
   Corresponding changes made in file engine-api/types/container/host_config.go

File: contrib/completion/bash/docker [hunk 3]
```diff
@@ -2379,6 +2379,8 @@ _docker_run() {
 		--volumes-from
 		--volume -v
 		--workdir -w
+		--net-classid
+		--interface-priority
```
Comment (albers) [inline contrib/completion/bash/docker]: please insert in alphabetical order.

**Comment by httpen2016** [comment]
changelist:
1. Insert in alphabetical order for bash completion.
2. Change `import path`, for code of `engine-api` is moved back into docker/docker repository

**Comment by FengtuWang** [comment]
changelist:
1. Change vendor's files as runc's version changed.
2. Add some codes to avoid setting cgroup when no `--net-classid` option specified.
In host which does't support net_cls, setting cgroup of net_cls will incure failure when start container.

**Comment by thaJeztah** [comment]
Not too familiar with these, but given that a container can be connected to multiple networks; is this something that should be on the _container_ or the _network_ (if possible somehow)?

**Comment by FengtuWang** [comment]
These two parameters can affect all networks in the container.
`--interface-priority` can assign different values to network interfaces in the container. 
For example:
`docker run -ti --interface-priority eth0:3 --interface-priority lo:5 ubuntu`

**Comment by thaJeztah** [comment]
> --interface-priority can assign different values to network interfaces in the container. 
> For example:
> docker run -ti --interface-priority eth0:3 --interface-priority lo:5 ubuntu

So, yes, wondering how that would work in combination with `docker network connect <network> <container>` / `docker network disconnect ...`

ping @mavenugo  - any thoughts?

**Comment by FengtuWang** [comment]
It can't work in combination with `docker network connect <network> <container> / docker network disconnect ...` currently. It seems that there are some problems in current modifies. I will try to fix it.

**Comment by thaJeztah** [comment]
ping @mavenugo @mrjana PTAL

**Comment by LK4D4** [comment]
ping @mrjana @mavenugo @aboch @sanimej

**Comment by mavenugo** [comment]
apologies on the delayed response. we need some time to think this through especially around more abstracted `network policy` concepts. There has been a lot of interest in network policy and having a generic mechanism at a service/application level will be more beneficial and prevent fragmented experience for containers.

**Comment by FengtuWang** [comment]
I want to remove the `--interface-priority` flag because it can't work. I can't figure out how to implement it, as net_prio of cgroup can't support network namespace currently.

**Comment by thaJeztah** [comment]
I'm going to close this PR for now, pending a proper design on how we want to handle these.

One thing brought up by @tianon, is that controlling outgoing network traffic could be more difficult as it (IIUC) does not use the container-container network

**Pull request closed without merging**